File: server/engines/tesseract.js
import { execFile } from 'child_process';

const TESSERACT_PATH = '/opt/homebrew/bin/tesseract';

export async function runTesseractOCR(imagePath, options = {}) {
  return new Promise((resolve, reject) => {
    const startTime = performance.now();
    const psm = options.psm || '3'; // Fully automatic page segmentation, but no OSD
    const lang = options.lang || 'eng';

    execFile(
      TESSERACT_PATH,
      [imagePath, 'stdout', '-l', lang, '--psm', psm],
      { maxBuffer: 10 * 1024 * 1024 },
      (error, stdout, stderr) => {
        const latencyMs = Math.round((performance.now() - startTime) * 10) / 10;
        if (error) {
          return reject(new Error(`Tesseract error: ${stderr || error.message}`));
        }

        const rawText = stdout.trim();
        const lines = rawText
          .split('\n')
          .map((line) => line.trim())
          .filter(Boolean)
          .map((text) => ({ text, confidence: 1.0 }));

        resolve({
          text: rawText,
          lines,
          latencyMs,
          engine: 'tesseract',
          engineName: 'Tesseract OCR (v5.5.1)'
        });
      }
    );
  });
}

File: server/engines/tesseract.py
import shutil
import subprocess
import time

TESSERACT_PATH = shutil.which("tesseract") or "/opt/homebrew/bin/tesseract"


def run_tesseract_ocr(image_path: str, options: dict = None) -> dict:
    if options is None:
        options = {}

    psm = str(options.get("psm", "3"))  # Fully automatic page segmentation, but no OSD
    lang = str(options.get("lang", "eng"))

    start_time = time.perf_counter()
    result = subprocess.run(
        [TESSERACT_PATH, str(image_path), "stdout", "-l", lang, "--psm", psm],
        capture_output=True,
        text=True,
        check=False
    )
    latency_ms = round((time.perf_counter() - start_time) * 1000, 1)

    if result.returncode != 0:
        raise RuntimeError(f"Tesseract error: {result.stderr or 'Process failed'}")

    raw_text = result.stdout.strip()
    lines = [
        {"text": line.strip(), "confidence": 1.0}
        for line in raw_text.splitlines()
        if line.strip()
    ]

    return {
        "text": raw_text,
        "lines": lines,
        "latencyMs": latency_ms,
        "engine": "tesseract",
        "engineName": "Tesseract OCR (v5.5.1)"
    }
